quickwit/quickwit-ingest/src/ingest_v2/replication.rs

```diff
@@ -0,0 +1,685 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::iter::once;
+use std::sync::Arc;
+
+use futures::StreamExt;
+use mrecordlog::MultiRecordLog;
+use quickwit_common::ServiceStream;
+use quickwit_proto::ingest::ingester::{
+    ack_replication_message, syn_replication_message, AckReplicationMessage, ReplicateRequest,
+    ReplicateResponse, ReplicateSuccess, SynReplicationMessage, TruncateRequest, TruncateResponse,
+};
+use quickwit_proto::ingest::{CommitTypeV2, IngestV2Error, IngestV2Result, ShardState};
+use quickwit_proto::types::NodeId;
+use tokio::sync::{mpsc, oneshot, watch, RwLock};
+use tokio::task::JoinHandle;
+
+use super::ingester::{commit_doc, IngesterState, Position, ReplicaShard, ShardStatus};
+use crate::metrics::INGEST_METRICS;
+
+#[derive(Debug)]
```

Comment:
/// A replication request is sent from the Leader to its follower to update the state of a queue.
```suggestion
#[derive(Debug)]
```